drivers/adc/ad9656/ad9656.c

```diff
@@ -0,0 +1,248 @@
+ * 	      User input data is supplied on the outputs instead of real data captured on the ADC.
+ * @param dev - The device handler for the ad9656 chip
+ * @param test_mode - The type of test that is to be performed or OFF if the testing
+ * 					  process is to be stopped
+ * @param user_input_test_pattern - User input test pattern that is sent on the output instead of
+ * 					the ADC	data
+ * @return SUCCESS if the ad9656 chip could be successfully set for user input testing, FAILURE
+ * 		   otherwise
+ */
+int32_t ad9656_user_input_test(struct ad9656_dev *dev, uint32_t test_mode,
+			       struct ad9656_user_input_test_pattern user_input_test_pattern)
+{
+	uint8_t format;
+
+	/* write user input test pattern in the registers */
+	ad9656_spi_write(dev, AD9656_REG_USER_TEST_PATTERN_1_LSB,
+			 user_input_test_pattern.user_test_pattern1 & 0xFF);
+	ad9656_spi_write(dev, AD9656_REG_USER_TEST_PATTERN_1_MSB,
+			 user_input_test_pattern.user_test_pattern1 >> 8);
+	ad9656_spi_write(dev, AD9656_REG_USER_TEST_PATTERN_2_LSB,
+			 user_input_test_pattern.user_test_pattern2 & 0xFF);
+	ad9656_spi_write(dev, AD9656_REG_USER_TEST_PATTERN_2_MSB,
+			 user_input_test_pattern.user_test_pattern2 >> 8);
+
+	/* determine the chip to output the user test pattern */
+	return ad9656_spi_write(dev, AD9656_REG_ADC_TEST_MODE, test_mode);
+}
+
+/**
+ * @brief Setup the working parameters of the ad9656 chip
+ * @param device - The device handler of the ad9656 chip
+ * @param init_param - Values for the working parameters of ad9656
+ * @return SUCCESS if device is ready for use, FAILURE otherwise
+ */
+int32_t ad9656_setup(struct ad9656_dev **device,
+		     const struct ad9656_init_param *init_param)
+{
+	uint8_t chip_id;
+	uint8_t pll_stat;
+	int32_t ret;
+	struct ad9656_dev *dev;
+
+	dev = (struct ad9656_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+
+	/* SPI */
+	ret = spi_init(&dev->spi_desc, &init_param->spi_init);
+	if (ret != SUCCESS)
+		return ret;
+
+	ad9656_spi_write(dev, AD9656_SPI_CONFIG, 0x3C);	/* RESET */
+	mdelay(250);
+
+	ret = ad9656_spi_read(dev, AD9656_REG_CHIP_ID, &chip_id);
+
+	if (ret != SUCCESS)
+		return ret;
+
+	if (chip_id != AD9656_CHIP_ID) {
+		printf("AD9656: Invalid CHIP ID (0x%x).\n", chip_id);
+		return FAILURE;
+	}
+
+	ad9656_spi_write(dev, AD9656_REG_LINK_CONTROL,
+			 0x15);	/* disable link, ilas enable */
+	ad9656_spi_write(dev, AD9656_REG_JESD204B_MF_CTRL,
+			 0x1f);	/* 32 frames per multiframe */
+	ad9656_spi_write(dev, AD9656_REG_JESD204B_M_CTRL, 0x03);	/* 4 converters */
+	ad9656_spi_write(dev, AD9656_REG_JESD204B_CSN_CONFIG,
+			 0x0d);	/* converter resolution of 14-bit */
+	ad9656_spi_write(dev, AD9656_REG_JESD204B_SUBCLASS_CONFIG,
+			 0x2f);	/* subclass-1, N'=16 */
+	ad9656_spi_write(dev, AD9656_REG_JESD204B_QUICK_CONFIG, 0x44);	/* m=4, l=4 */
+	ad9656_spi_write(dev, AD9656_REG_JESD204B_SCR_L,
+			 0x83);	/* enable scrambling, l=4 */
+	if (init_param->lane_rate_kbps < 2000000)
```

Comment:
this can be restructured by moving the spi_write outside of the if statement and using a tmp variable to enable or disable the line rate. Remeber to check the return value of spi_write

Reply:
Done in v6